naas/requests/campaigns.py

```diff
@@ -0,0 +1,84 @@
+from naas.client import Client
+
+
+class Campaigns:
+
+    @staticmethod
+    def list_by_project_id(project_id, params=None):
+        """
+        Retrieve the list of campaigns by project
+        :param project_id: string
+        :param params: dict
+        :return: Response
+        """
+        if params is None:
+            params = {}
+        route = Client.create_route('rels/project-campaigns')
```

Comment:
@muhallan - an important note here is that the `rels/*` is the [IANA Link Relationship](https://www.iana.org/assignments/link-relations/link-relations.xhtml). We use official `rel`s where applicable, then we use a fully qualified URI as our custom (per spec on defining custom link relationships). This is _not_ the actual URL that gets requested, but is the constant string we use to lookup the URL in our directory. For instance, here is one item returned from the **Directory**:

```json
{
   "title":"Project Campaigns",
   "href":"https://naas-api-staging.tinylab.com/projects/{project_id}/campaigns{?page,per_page}",
   "rel":"https://naas-api-staging.tinylab.com/rels/project-campaigns",
   "templated":true
}
```

The API client should never have to manually build URLs. The client will always lookup the `Link` item and then _expand_ the url as needed. An example from `ruby`:

```ruby
project_id = 'project_12234'

routes = Naas::Client.routes
route  = routes.find_by_rel('https://naas-api-staging.tinylab.com/rels/project-campaigns')
url    = route.url_for(id: project_id, page: 2, per_page: 100)
#=> https://naas-api-staging.tinylab.com/projects/project_12234/campaigns?page=2&per_page=100
```

Once you have resolved the `url` from the lookup, then you can use that URL to make the _request_ itself.

Reply:
Thank you. I have noticed that I need to add more methods to the `Client` class. Let me work on adding the `rel_for`, `routes`, `directory`, e.t.c. methods since they are required here.